File: desktop/backend/src/diveintocrypto_desktop/scan/microstructure.py
"""Futures-microstructure overlay — a scan-layer signal bundle over the
already-assembled ``series_data`` (OI / funding / taker / L-S). Companion to
``scan/divergence.py``: it does NOT touch the parity-locked 15-indicator consensus
or the whale-divergence verdict. It is purely additive — ``assemble()`` attaches its
result to the per-symbol contract as a ``microstructure`` field for the UI/analytics.

Every signal is a pure function over aligned float lists, strictly causal (reads only
the closed window), and returns a value in [-1, +1] (+ bullish / − bearish) or None
when there is not enough data. The bundle score is the weighted mean of the available
signals, mapped to a signed −100..+100 and a BUY/SELL/NEUTRAL label.

Data (from ``symbol_builder.build_symbol``; ~48 points at 5m cadence):
  oi      top-of-window open interest (base contracts)
  funding funding-rate history (fraction, e.g. 0.0001)
  taker   taker buy/sell ratio (≈1.0 balanced; >1 = buy aggression)
  glob    global account long/short ratio (retail / "dumb money")
  pos     top-trader position long/short ratio ("smart money")
  price   5m close series
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Params:
    window: int = 24          # bars considered for trend/z-score
    min_points: int = 8       # minimum series length to compute a signal
    z_cap: float = 3.0        # z-scores clamped to +/- this before scaling
    buy_threshold: float = 20.0   # |score| >= -> directional label
    strong_threshold: float = 55.0
    # per-signal blend weights (only the available ones are used)
    w_oi_div: float = 1.0
    w_oi_breakout: float = 1.0
    w_funding_fade: float = 1.0
    w_taker: float = 1.0
    w_crowding: float = 1.0
    w_smart_dumb: float = 0.8


# ── small numeric helpers (stdlib only) ──────────────────────────────────────
def _clean(x: list[float] | None) -> list[float]:
    if not x:
        return []
    out: list[float] = []
    for v in x:
        try:
            f = float(v)
        except (TypeError, ValueError):
            continue
        if f == f and f not in (float("inf"), float("-inf")):  # drop NaN/Inf
            out.append(f)
    return out


def _mean(a: list[float]) -> float:
    return sum(a) / len(a) if a else 0.0


def _std(a: list[float]) -> float:
    n = len(a)
    if n < 2:
        return 0.0
    m = _mean(a)
    return (sum((v - m) ** 2 for v in a) / (n - 1)) ** 0.5


def _clip(x: float, lo: float = -1.0, hi: float = 1.0) -> float:
    return lo if x < lo else hi if x > hi else x


def _sign(x: float) -> int:
    return 1 if x > 0 else -1 if x < 0 else 0


def _pct_change(a: list[float]) -> float:
    """Signed relative change first→last over the window; 0 if degenerate."""
    if len(a) < 2 or a[0] == 0.0:
        return 0.0
    return (a[-1] - a[0]) / abs(a[0])


def _zscore(a: list[float], cap: float) -> float:
    """Z-score of the last point vs the preceding window; clamped to +/- cap."""
    if len(a) < 3:
        return 0.0
    hist = a[:-1]
    s = _std(hist)
    if s == 0.0:
        return 0.0
    return _clip((a[-1] - _mean(hist)) / s, -cap, cap)


# ── individual signals: each returns (value in [-1,1] | None, reason) ─────────
def _oi_price_divergence(oi: list[float], price: list[float], p: Params):
    n = min(len(oi), len(price))
    if n < p.min_points:
        return None, "oi/price insufficient"
    oi_w, pr_w = oi[-p.window:], price[-p.window:]
    d_oi, d_pr = _pct_change(oi_w), _pct_change(pr_w)
    if abs(d_pr) < 0.002:
        return 0.0, "price flat"
    # OI rising WITH the move = real money confirms trend (+ in trend direction);
    # OI falling while price moves = short-covering / weak move -> fade the move.
    confirm = 1.0 if d_oi > 0 else -1.0
    mag = _clip(abs(d_pr) / 0.03)
    return _clip(_sign(d_pr) * confirm * mag), f"dPr={d_pr:+.3f} dOI={d_oi:+.3f}"


def _oi_breakout_confirm(oi: list[float], price: list[float], p: Params):
    n = min(len(oi), len(price))
    if n < p.min_points:
        return None, "oi/price insufficient"
    oi_w, pr_w = oi[-p.window:], price[-p.window:]
    d_oi, d_pr = _pct_change(oi_w), _pct_change(pr_w)
    if abs(d_pr) < 0.01 or d_oi <= 0:  # only fires on a real move backed by OI expansion
        return 0.0, "no OI-backed breakout"
    return _clip(_sign(d_pr) * _clip(d_oi / 0.05)), f"breakout dPr={d_pr:+.3f} dOI={d_oi:+.3f}"


def _funding_fade(funding: list[float], p: Params):
    f = funding[-p.window:]
    if len(f) < 3:
        return None, "funding insufficient"
    z = _zscore(f, p.z_cap)
    if z == 0.0:
        return 0.0, "funding neutral"
    # Extreme positive funding = crowded longs pay shorts -> contrarian bearish.
    return _clip(-z / p.z_cap), f"funding z={z:+.2f}"


def _taker_aggression(taker: list[float], p: Params):
    t = taker[-p.window:]
    if len(t) < p.min_points:
        return None, "taker insufficient"
    recent = _mean(t[-4:]) if len(t) >= 4 else _mean(t)
    level = _clip((recent - 1.0) / 0.30)          # >1 buy pressure, <1 sell pressure
    trend = _clip(_pct_change(t) / 0.10)
    return _clip(0.6 * level + 0.4 * trend), f"taker~{recent:.3f}"


def _ls_crowding_fade(glob: list[float], p: Params):
    g = glob[-p.window:]
    if len(g) < 3:
        return None, "glob insufficient"
    z = _zscore(g, p.z_cap)
    if z == 0.0:
        return 0.0, "crowd neutral"
    # Retail account L/S extreme = crowded -> contrarian.
    return _clip(-z / p.z_cap), f"crowd z={z:+.2f}"


def _smart_dumb_spread(pos: list[float], glob: list[float], p: Params):
    n = min(len(pos), len(glob))
    if n < p.min_points:
        return None, "pos/glob insufficient"
    spread = _mean(pos[-p.window:]) - _mean(glob[-p.window:])
    if abs(spread) < 0.02:
        return 0.0, "spread flat"
    # Smart money (top-trader positions) more long than retail accounts -> bullish.
    return _clip(spread / 0.5), f"smart-dumb={spread:+.3f}"


_SIGNALS = (
    ("oi_price_divergence", "w_oi_div"),
    ("oi_breakout_confirm", "w_oi_breakout"),
    ("funding_fade", "w_funding_fade"),
    ("taker_aggression", "w_taker"),
    ("ls_crowding_fade", "w_crowding"),
    ("smart_dumb_spread", "w_smart_dumb"),
)


def evaluate(series_data: dict, enabled: bool = True, params: Params = Params()) -> dict:
    """Run the microstructure signal bundle over ``series_data``.

    Returns ``{score, direction, label, active, signals:[{name,score,reason,weight}]}``.
    ``score`` is a signed −100..+100 (+bullish/−bearish); ``active`` = signals with data.
    Safe on missing/short series (those signals are skipped). ``enabled=False`` short-circuits.
    """
    if not enabled:
        return {"score": 0.0, "direction": 0, "label": "OFF", "active": 0, "signals": []}

    oi = _clean(series_data.get("oi"))
    price = _clean(series_data.get("price"))
    funding = _clean(series_data.get("funding"))
    taker = _clean(series_data.get("taker"))
    glob = _clean(series_data.get("glob"))
    pos = _clean(series_data.get("pos"))

    computed = {
        "oi_price_divergence": _oi_price_divergence(oi, price, params),
        "oi_breakout_confirm": _oi_breakout_confirm(oi, price, params),
        "funding_fade": _funding_fade(funding, params),
        "taker_aggression": _taker_aggression(taker, params),
        "ls_crowding_fade": _ls_crowding_fade(glob, params),
        "smart_dumb_spread": _smart_dumb_spread(pos, glob, params),
    }

    signals: list[dict] = []
    wsum = 0.0
    acc = 0.0
    for name, wattr in _SIGNALS:
        val, reason = computed[name]
        if val is None:
            continue
        w = getattr(params, wattr)
        signals.append({"name": name, "score": round(val, 4), "reason": reason, "weight": w})
        acc += val * w
        wsum += w

    if wsum == 0.0:
        return {"score": 0.0, "direction": 0, "label": "NEUTRAL", "active": 0, "signals": []}

    score = round(_clip(acc / wsum) * 100.0, 1)
    mag = abs(score)
    if mag < params.buy_threshold:
        label = "NEUTRAL"
        direction = 0
    else:
        direction = _sign(score)
        strong = mag >= params.strong_threshold
        label = ("STRONG_BUY" if strong else "BUY") if direction > 0 else ("STRONG_SELL" if strong else "SELL")

    return {"score": score, "direction": direction, "label": label, "active": len(signals), "signals": signals}

File: android/app/src/commonMain/kotlin/com/diveintocrypto/android/domain/overlay/Microstructure.kt
package com.diveintocrypto.android.domain.overlay

import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Futures-microstructure overlay — a scan-layer signal bundle over the assembled
 * series data (OI / funding / taker / L-S). Faithful port of the Python reference
 * `scan/microstructure.py`.
 *
 * Every signal is a pure function over aligned float lists, strictly causal, and
 * returns a value in [-1, +1] (+ bullish / − bearish) or null when there is not
 * enough data. The bundle score is the weighted mean of the available signals,
 * mapped to a signed −100..+100 and a BUY/SELL/NEUTRAL label. ADDITIVE — it never
 * touches the parity-locked consensus.
 */
object Microstructure {

    data class Params(
        val window: Int = 24,
        val minPoints: Int = 8,
        val zCap: Double = 3.0,
        val buyThreshold: Double = 20.0,
        val strongThreshold: Double = 55.0,
        val wOiDiv: Double = 1.0,
        val wOiBreakout: Double = 1.0,
        val wFundingFade: Double = 1.0,
        val wTaker: Double = 1.0,
        val wCrowding: Double = 1.0,
        val wSmartDumb: Double = 0.8,
    )

    data class SignalOut(val name: String, val score: Double, val reason: String, val weight: Double)

    data class Result(
        val score: Double,      // signed −100..+100
        val direction: Int,     // +1 / −1 / 0
        val label: String,      // STRONG_BUY | BUY | NEUTRAL | SELL | STRONG_SELL | OFF
        val active: Int,
        val signals: List<SignalOut>,
    )

    data class Series(
        val oi: List<Double> = emptyList(),
        val price: List<Double> = emptyList(),
        val funding: List<Double> = emptyList(),
        val taker: List<Double> = emptyList(),
        val glob: List<Double> = emptyList(),
        val pos: List<Double> = emptyList(),
    )

    // ── numeric helpers (mirror the Python file) ────────────────────────────
    private fun clean(x: List<Double>): List<Double> =
        x.filter { !it.isNaN() && !it.isInfinite() }

    private fun mean(a: List<Double>): Double = if (a.isEmpty()) 0.0 else a.sum() / a.size

    private fun std(a: List<Double>): Double {
        val n = a.size
        if (n < 2) return 0.0
        val m = mean(a)
        return sqrt(a.sumOf { (it - m) * (it - m) } / (n - 1))
    }

    private fun clip(x: Double, lo: Double = -1.0, hi: Double = 1.0): Double =
        if (x < lo) lo else if (x > hi) hi else x

    private fun sign(x: Double): Int = if (x > 0) 1 else if (x < 0) -1 else 0

    private fun pctChange(a: List<Double>): Double {
        if (a.size < 2 || a.first() == 0.0) return 0.0
        return (a.last() - a.first()) / abs(a.first())
    }

    private fun zscore(a: List<Double>, cap: Double): Double {
        if (a.size < 3) return 0.0
        val hist = a.subList(0, a.size - 1)
        val s = std(hist)
        if (s == 0.0) return 0.0
        return clip((a.last() - mean(hist)) / s, -cap, cap)
    }

    private fun List<Double>.tail(n: Int): List<Double> =
        if (size <= n) this else subList(size - n, size)

    // ── individual signals: (value in [-1,1] | null, reason) ────────────────
    private fun oiPriceDivergence(oi: List<Double>, price: List<Double>, p: Params): Pair<Double?, String> {
        if (minOf(oi.size, price.size) < p.minPoints) return null to "oi/price insufficient"
        val dOi = pctChange(oi.tail(p.window))
        val dPr = pctChange(price.tail(p.window))
        if (abs(dPr) < 0.002) return 0.0 to "price flat"
        val confirm = if (dOi > 0) 1.0 else -1.0
        val mag = clip(abs(dPr) / 0.03)
        return clip(sign(dPr) * confirm * mag) to "dPr=$dPr dOI=$dOi"
    }

    private fun oiBreakoutConfirm(oi: List<Double>, price: List<Double>, p: Params): Pair<Double?, String> {
        if (minOf(oi.size, price.size) < p.minPoints) return null to "oi/price insufficient"
        val dOi = pctChange(oi.tail(p.window))
        val dPr = pctChange(price.tail(p.window))
        if (abs(dPr) < 0.01 || dOi <= 0) return 0.0 to "no OI-backed breakout"
        return clip(sign(dPr) * clip(dOi / 0.05)) to "breakout dPr=$dPr dOI=$dOi"
    }

    private fun fundingFade(funding: List<Double>, p: Params): Pair<Double?, String> {
        val f = funding.tail(p.window)
        if (f.size < 3) return null to "funding insufficient"
        val z = zscore(f, p.zCap)
        if (z == 0.0) return 0.0 to "funding neutral"
        return clip(-z / p.zCap) to "funding z=$z"
    }

    private fun takerAggression(taker: List<Double>, p: Params): Pair<Double?, String> {
        val t = taker.tail(p.window)
        if (t.size < p.minPoints) return null to "taker insufficient"
        val recent = if (t.size >= 4) mean(t.tail(4)) else mean(t)
        val level = clip((recent - 1.0) / 0.30)
        val trend = clip(pctChange(t) / 0.10)
        return clip(0.6 * level + 0.4 * trend) to "taker~$recent"
    }

    private fun lsCrowdingFade(glob: List<Double>, p: Params): Pair<Double?, String> {
        val g = glob.tail(p.window)
        if (g.size < 3) return null to "glob insufficient"
        val z = zscore(g, p.zCap)
        if (z == 0.0) return 0.0 to "crowd neutral"
        return clip(-z / p.zCap) to "crowd z=$z"
    }

    private fun smartDumbSpread(pos: List<Double>, glob: List<Double>, p: Params): Pair<Double?, String> {
        if (minOf(pos.size, glob.size) < p.minPoints) return null to "pos/glob insufficient"
        val spread = mean(pos.tail(p.window)) - mean(glob.tail(p.window))
        if (abs(spread) < 0.02) return 0.0 to "spread flat"
        return clip(spread / 0.5) to "smart-dumb=$spread"
    }

    fun evaluate(series: Series, enabled: Boolean = true, params: Params = Params()): Result {
        if (!enabled) return Result(0.0, 0, "OFF", 0, emptyList())

        val oi = clean(series.oi); val price = clean(series.price)
        val funding = clean(series.funding); val taker = clean(series.taker)
        val glob = clean(series.glob); val pos = clean(series.pos)

        val computed = listOf(
            Triple("oi_price_divergence", oiPriceDivergence(oi, price, params), params.wOiDiv),
            Triple("oi_breakout_confirm", oiBreakoutConfirm(oi, price, params), params.wOiBreakout),
            Triple("funding_fade", fundingFade(funding, params), params.wFundingFade),
            Triple("taker_aggression", takerAggression(taker, params), params.wTaker),
            Triple("ls_crowding_fade", lsCrowdingFade(glob, params), params.wCrowding),
            Triple("smart_dumb_spread", smartDumbSpread(pos, glob, params), params.wSmartDumb),
        )

        val signals = ArrayList<SignalOut>()
        var wsum = 0.0
        var acc = 0.0
        for ((name, valueReason, w) in computed) {
            val (value, reason) = valueReason
            if (value == null) continue
            signals += SignalOut(name, value, reason, w)
            acc += value * w
            wsum += w
        }
        if (wsum == 0.0) return Result(0.0, 0, "NEUTRAL", 0, emptyList())

        val score = kotlin.math.round(clip(acc / wsum) * 1000.0) / 10.0
        val mag = abs(score)
        val direction: Int
        val label: String
        if (mag < params.buyThreshold) {
            direction = 0; label = "NEUTRAL"
        } else {
            direction = sign(score)
            val strong = mag >= params.strongThreshold
            label = if (direction > 0) (if (strong) "STRONG_BUY" else "BUY")
                    else (if (strong) "STRONG_SELL" else "SELL")
        }
        return Result(score, direction, label, signals.size, signals)
    }
}
